File: java/146.java
import java.util.HashMap;

class LRUCache {
    class Node {
        int key;
        int value;
        Node pre;
        Node next;
    }
    private HashMap<Integer, Node> cache = new HashMap<>();
    private int size;
    private int capacity;
    private Node head;
    private Node tail;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.pre = head;
    }
    private void addNode(Node node) {
        node.pre = head;
        node.next = head.next;

        head.next = node;
        node.next.pre = node;
    }
    private void moveToHead(Node node) {
        removeNode(node);
        addNode(node);
    }
    private void removeNode(Node node) {
        node.pre.next = node.next;
        node.next.pre = node.pre;
    }
    private Node popTail() {
        Node node = tail.pre;
        removeNode(node);
        return node;
    }
    public int get(int key) {
        Node node = cache.get(key);
        if (node == null)
            return -1;
        moveToHead(node);
        return node.value;
    }

    public void put(int key, int value) {
        Node node = cache.get(key);
        if (node != null) {
            node.value = value;
            moveToHead(node);
        } else {
            Node new_node = new Node();
            new_node.key = key;
            new_node.value = value;
            if (size < capacity) {
                cache.put(key, new_node);
                addNode(new_node);
                size++;
            } else {
                Node remove_node = popTail();
                cache.remove(remove_node.key);
                addNode(new_node);
                cache.put(key, new_node);
                size++;
            }
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: python/146.py
class DLinkedNode():
    def __init__(self):
        self.key = 0
        self.value = 0
        self.pre = None
        self.next = None


class LRUCache():
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.head = DLinkedNode()
        self.tail = DLinkedNode()
        self.head.next = self.tail
        self.tail.pre = self.head
        self.dit = {}

    def _addNode(self, new_node):
        next = self.head.next

        self.head.next = new_node
        new_node.pre = self.head
        new_node.next = next
        next.pre = new_node

    def _removeNode(self, delete_node):
        pre = delete_node.pre
        next = delete_node.next
        pre.next = next
        next.pre = pre

    def _move_to_head(self, node):
        self._removeNode(node)
        self._addNode(node)

    def _delete_tail_node(self):
        node = self.tail.pre
        self._removeNode(node)
        return node

    def get(self, key: int) -> int:
        if key in self.dit:
            node = self.dit[key]
            self._move_to_head(node)
            return node
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.dit:
            node = self.dit[key]
            node.value = value
            self._move_to_head(node)
        else:
            new_node = DLinkedNode()
            new_node.key = key
            new_node.value = value
            if len(self.dit) >= self.capacity:
                deleted_node = self._delete_tail_node()
                self.dit.pop(deleted_node.key)
            self._addNode(new_node)
            self.dit[key] = new_node

n = DLinkedNode()
n.key = 1
n.value = 1